### PR TITLE
[dune] [build] Enable native compilation for the standard library.

### LINKED ISSUE #8860: Dune should not disable `native_compute`
I realized recently that the Dune build disables `native_compute`. I'm opening this issue to remember this should be fixed before we switch to Dune.

I understand the way `native_compute` artifacts are generated should probably be refined to better fit the Dune conventions. I'd be happy to help with that, although it is not clear to me what can be done given the limited interfacing we have with the OCaml compiler.

### PR DESCRIPTION
**Pull request by ejgallego**
(Mostly) fixes #8860 ; see documentation for more information.

There are some open issues with regards to our CI + dune, so far we
build everything in dev mode, except in the `pkg:opam` job; I wonder
whether we should keep doing that.

This PR also closes #13641 by side-effect as it pulls dune 2.8 which
fixes that issue.

### CODE REVIEW

File: dune-project
Comment (SkySkimmer) [inline dune-project:1]: Is this in opam yet?

Comment (ejgallego) [reply dune-project:1]: Not yet, hence the tag.

File: dev/doc/build-system.dune.md
Comment (Zimmi48) [inline dev/doc/build-system.dune.md:183]: Is this a general feature of Dune native compilation support?

Comment (ejgallego) [reply dev/doc/build-system.dune.md:183]: What do you mean by general @Zimmi48 ? For now the rule governing this hardcoded but actually it would be pretty easy to do add a per-profile setting.

Comment (Zimmi48) [reply dev/doc/build-system.dune.md:183]: I wanted to know if this was hard-coded in Dune because I couldn't see what could be triggering this in the code of Coq.

Comment (ejgallego) [reply dev/doc/build-system.dune.md:183]: Yes this is hardcoded in Dune, when in `dev` profile.

**Comment by gares** [comment]
@ejgallego this PR is in the 8.13+beta1 milestone. Can you update us on the status of the external fix? Is it good now?

**Comment by Zimmi48** [comment]
AFAIK this is waiting on the merge of https://github.com/ocaml/dune/pull/3210 and the release of 2.8. I don't think there is any issue to delaying this to 8.14.

**Comment by ejgallego** [comment]
Indeed no harm in delaying, tho the dune upstream can now be merged as 8.12. has been released.

Main advantage of having this in 8.13 would be that the pinned opam packages would include the native libs, so indeed if we want to backport the PR is small enough [and it doesn't touch critical parts] so it should be easy to do.

**Comment by ejgallego** [comment]
@SkySkimmer @Zimmi48 this is ready for review; not ideal yet due to lack of configurability but it does mean that opam packages with native are much easier to do.

**Comment by gares** [comment]
Is the choice `dev -> no native` hardcoded in dune?

**Comment by ejgallego** [comment]
> Is the choice `dev -> no native` hardcoded in dune?

Yup, but would be easy to have it configurable once we see the use cases.

Hopefully we will move to #13287 soon so this is a transient status anyways.

**Comment by gares** [comment]
This patch makes dune projects not fail on warnings as a side effect of this design choice.

**Comment by ejgallego** [comment]
> This patch makes dune projects not fail on warnings as a side effect of this design choice.

Good point, we used to run check in the linter so indeed that could be a solution.

**Review by Zimmi48 — APPROVED** [review]

**Comment by Zimmi48** [comment]
>Good point, we used to run check in the linter so indeed that could be a solution.

Used to? It is still the case. We call [`dev/lint-commits.sh`](https://github.com/coq/coq/blob/master/dev/lint-commits.sh) to make sure all commits in a PR build (without warnings).

**Comment by Zimmi48** [comment]
I propose to merge this PR on Wednesday if there are no more comments.

**Comment by ejgallego** [comment]
> Used to? It is still the case. We call [`dev/lint-commits.sh`](https://github.com/coq/coq/blob/master/dev/lint-commits.sh) to make sure all commits in a PR build (without warnings).

Glad to hear, I didn't check.

So with this configuration we will allow builds with warning to proceed so that may not be a bad thing.

Anyways I see many possible different combinations we'd like to try, but my feeling is that it is more correct to use `--profile=release` in the CI at least for jobs that would test against an "installed" Coq, but if we want some particular config we can just try.

As of today only `dev` will skip native build which is IMHO what you want when developing Coq, but we could make that configurable. I didn't do it just because it seems it is going to be a short-lived tweak anyways if for 8.14 we have a split setup.

**Comment by Zimmi48** [comment]
>So with this configuration we will allow builds with warning to proceed so that may not be a bad thing.

Interesting! The lint CI job will report the warnings, but we will still get the results of testing the jobs that depend on the main Dune job.

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml:122]:
```suggestion
    - dune build -p coq,coqide-server,coqide @install
```
not having a space here looks terrible imo

Comment (ejgallego) [reply .gitlab-ci.yml]: Done.

**Comment by Zimmi48** [comment]
Actually, since this PR bumps the minimal version of Dune, merging it now would make merging #13617 more difficult (since @JasonGross needs to be able to update his Ubuntu ppa for CI purposes and Dune 2.8 is not yet packaged on Debian / Ubuntu).

**Comment by ejgallego** [comment]
> Actually, since this PR bumps the minimal version of Dune, merging it now would make merging #13617 more difficult (since @JasonGross needs to be able to update his Ubuntu ppa for CI purposes and Dune 2.8 is not yet packaged on Debian / Ubuntu).

Dune 2.8 will land in Debian soon so I wouldn't be too worried.

**Comment by ejgallego** [comment]
Demilestoning, this is waiting on the PR adding coqnative support to dune now.

**Pull request closed without merging**
